### PR TITLE
Remove Livedata usage for events in the home activity 

### PR DESCRIPTION
**Pull request by GabrielFleischer**
This PR removes the usage of live data when dispatching events.

Currently, the logic on how to switch to a fragment/activity is held in the view model.
This is not the best way to do it and should be addressed later.

### CODE REVIEW

**Comment by Warhedgehog** [comment]
I've not yet looked in depth but you might want to fix the little code smell

**Comment by pierluca** [comment]
About that code smell, shouldn't we just switch that function to use a primitive boolean, instead of `java.lang.Boolean` ?

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeActivity.java [hunk 1]
```diff
@@ -405,28 +165,45 @@ private void handleSocialMediaNavigation() {
...
   }
 
+  public void requestCameraPermission() {
+    // Setup result listener to open the connect tab once the permission is granted
+    getSupportFragmentManager()
+        .setFragmentResultListener(
+            CameraPermissionFragment.REQUEST_KEY, this, (k, b) -> openConnect());
+
+    mViewModel.openCameraPermission(getSupportFragmentManager(), getActivityResultRegistry());
+  }
```
Comment (pierluca) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeActivity.java:205]: what happens if the camera permission is denied ?
I don't see a code path to deal with that situation / log it / warn the user in some way.

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeViewModel.java [hunk 2]
```diff
@@ -162,34 +147,25 @@ protected void onCleared() {
...
-                      .subscribe(lao.getChannel())
-                      .subscribe(
-                          () -> {
-                            Log.d(TAG, "subscribing to LAO with id " + lao.getId());
-                            openLAO(lao.getId());
-                          },
-                          error ->
-                              ErrorUtils.logAndShow(
-                                  getApplication(), TAG, error, R.string.error_create_lao));
+                  Log.d(TAG, "subscribing to LAO with id " + lao.getId());
```
Comment (pierluca) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeViewModel.java:167]: once this is called, we're already subscribed, right ?

**Review by pierluca — COMMENTED** [review]
Looks good, barring minor details.
I'm not too familiar with all the operations surrounding fragments though, @Warhedgehog will have a more informed opinion on that :-)

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeActivity.java (hunk 1 above)
Comment (GabrielFleischer) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeActivity.java:205]: It will reopen Connect which will redo the permission check and if the test fails it will re-ask the user.

It is basically a loop.

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeViewModel.java (hunk 2 above)
Comment (GabrielFleischer) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeViewModel.java:167]: Yes, good point

Comment (GabrielFleischer) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeViewModel.java]: Fixed

**Comment by GabrielFleischer** [comment]
> About that code smell, shouldn't we just switch that function to use a primitive boolean, instead of `java.lang.Boolean` ?

The thing is, LiveData produces Booleans which are Objects and can therefore be null.

I made a change that addresses the code smell.

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeActivity.java (hunk 1 above)
Comment (Warhedgehog) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeActivity.java:205]: Upon testing, I noticed that when denying, we go to the home fragment. In itself, it is not a problem but after twice denying, when clicking on the allow button (ours) it does not ask for permission but goes back to home like I denied permission

Moreover, I think the behaviour should be commented and if possible the logic behind. Imo it is not immediately clear from the code.

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeViewModel.java [hunk 3]
```diff
@@ -199,147 +170,112 @@ public void launchLao() {
...
-    mOpenHomeEvent.postValue(new SingleEvent<>(true));
+  public void openLaunch(FragmentManager manager) {
+    setCurrentFragment(manager, R.id.fragment_launch, LaunchFragment::newInstance);
   }
 
   public void openConnecting(String laoId) {
-    mOpenConnectingEvent.postValue(new SingleEvent<>(laoId));
+    Intent intent = new Intent(getApplication().getApplicationContext(), ConnectingActivity.class);
+    intent.putExtra(Constants.LAO_ID_EXTRA, laoId);
+    getApplication().startActivity(intent);
```
Comment (Warhedgehog) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeViewModel.java:210]: Did you forget to pass the activity in the arguments like the other activity start ?

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeActivity.java [hunk 4]
```diff
@@ -279,7 +129,9 @@ public boolean onCreateOptionsMenu(Menu menu) {
   @Override
   public boolean onOptionsItemSelected(MenuItem item) {
     if (item.getItemId() == R.id.settings) {
-      mViewModel.openSettings();
+      Intent intent = new Intent(this, SettingsActivity.class);
+      Log.d(HomeViewModel.TAG, "Trying to open settings");
+      startActivity(intent);
```
Comment (Warhedgehog) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeActivity.java:134]: Most of the activity starting is done in ViewModel, is there a reason to do it here ?

**Review by Warhedgehog — CHANGES_REQUESTED** [review]
Overall massive improvement of course. Couple things I'd like addressed 

- The new implementation of camera permission has a bug I think (double deny)
- There is some fragment and activity navigation in both the viewModel and the activity, should we not stick with just 1 ? Also some of the activity navigation methods are static and some not, would it not be better to be consistent ?

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeActivity.java [hunk 5]
```diff
@@ -226,25 +89,12 @@ private void subscribeWalletEvents() {
   @Override
   protected void onResume() {
     super.onResume();
-    mViewModel.openHome();
+    mViewModel.openHome(getSupportFragmentManager());
```
[the remark this answers is not in the document]
Comment (GabrielFleischer) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeActivity.java:92]: It would be more impactful and, to me, go outside the scope of the PR.

We could do a PR to move the activity and fragment start in their respective class and then apply the livedata change.

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeViewModel.java (hunk 3 above)
Comment (GabrielFleischer) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeViewModel.java:210]: Yes 😅

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeActivity.java (hunk 4 above)
Comment (GabrielFleischer) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeActivity.java:134]: I missed this one

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeActivity.java [hunk 6]
```diff
@@ -55,7 +46,7 @@ protected void onCreate(Bundle savedInstanceState) {
     super.onCreate(savedInstanceState);
     setContentView(R.layout.home_activity);
 
-    setupHomeFragment();
+    setCurrentFragment(getSupportFragmentManager(), R.id.fragment_home, HomeFragment::newInstance);
```
[the remark this answers is not in the document]
Comment (GabrielFleischer) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeActivity.java:55]: I moved everything wher the call was made. That way the reader knows what is happening where the call is made.

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeViewModel.java [hunk 7]
```diff
@@ -92,20 +74,14 @@ public HomeViewModel(
...
 
     mLAOs =
         LiveDataReactiveStreams.fromPublisher(
-            this.laoRepository.getAllLaos().toFlowable(BackpressureStrategy.BUFFER));
-  }
-
-  @Override
-  public void onPermissionGranted() {
-    openQrCodeScanning();
+            laoRepository.getAllLaos().toFlowable(BackpressureStrategy.BUFFER));
```
Comment (Warhedgehog) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeViewModel.java:84]: Why this change ?

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeViewModel.java [hunk 8]
```diff
@@ -199,147 +166,44 @@ public void launchLao() {
...
-  }
-
   public void setLaoName(String name) {
     this.mLaoName.setValue(name);
   }
 
-  public void setIsWalletSetUp(Boolean isSetUp) {
+  public void setIsWalletSetUp(boolean isSetUp) {
     this.mIsWalletSetUp.setValue(isSetUp);
   }
```
Comment (Warhedgehog) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeViewModel.java:190]: I know this is not from you and not the point of the PR but would it not be better to use postValue ?

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeViewModel.java [hunk 9]
```diff
@@ -199,147 +166,44 @@ public void launchLao() {
...
     wallet.logout();
     setIsWalletSetUp(false);
-    openWallet();
+  }
+
+  public void openConnecting(String laoId) {
+    Intent intent = new Intent(getApplication().getApplicationContext(), ConnectingActivity.class);
+    intent.putExtra(Constants.LAO_ID_EXTRA, laoId);
+    intent.addFlags(FLAG_ACTIVITY_NEW_TASK);
+    getApplication().startActivity(intent);
```
Comment (Warhedgehog) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeViewModel.java:207]: I'm not too fan about having some activity navigation in the viewModel but I have no better idea

**Review by Warhedgehog — APPROVED** [review]
LGTM. Couple questions/remark with which you do what you want :)

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeViewModel.java (hunk 9 above)
Comment (GabrielFleischer) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeViewModel.java:207]: With the way the QR Scanner is done, there is no clean way to move this. So for now it will stay here.

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeViewModel.java (hunk 8 above)
Comment (GabrielFleischer) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeViewModel.java:190]: Maybe, I am not certain of what the actual impact would be
